File: backend/app/shared/period.py
"""Período operacional compartilhado por todos os módulos.

Porte literal de `src/lib/period.ts` (Next.js). Único ponto de verdade para a
classificação de uma competência mensal (ano-calendário + mês) no semestre
operacional da empresa. NÃO reimplementar esta regra em outro lugar.

Regra oficial:
  - dezembro pertence ao S1 do ano SEGUINTE (ex.: Dez/2026 -> 2027 S1);
  - janeiro a maio pertencem ao S1 do próprio ano (ex.: Mar/2027 -> 2027 S1);
  - junho a novembro pertencem ao S2 do próprio ano (ex.: Ago/2027 -> 2027 S2).

Para S1, o "ano do período" é sempre o ano de TÉRMINO (maio), nunca o de
início (dezembro do ano anterior). Ex.: cycle_from_year_semester(2027, "S1")
= Dez/2026 - Mai/2027.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from enum import Enum
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from sqlalchemy.sql.elements import ColumnElement

MONTH_NAMES = [
    "Jan", "Fev", "Mar", "Abr", "Mai", "Jun",
    "Jul", "Ago", "Set", "Out", "Nov", "Dez",
]

MONTH_NAMES_FULL = [
    "Janeiro", "Fevereiro", "Março", "Abril", "Maio", "Junho",
    "Julho", "Agosto", "Setembro", "Outubro", "Novembro", "Dezembro",
]


class Semester(str, Enum):
    S1 = "S1"
    S2 = "S2"


@dataclass(frozen=True, slots=True)
class PeriodRange:
    start_year: int
    start_month: int  # 1-12
    end_year: int
    end_month: int  # 1-12


@dataclass(frozen=True, slots=True)
class MonthReference:
    year: int
    month: int


@dataclass(frozen=True, slots=True)
class OperationalPeriod:
    period_year: int
    semester: Semester


def to_month_index(year: int, month: int) -> int:
    return year * 12 + (month - 1)


def _from_month_index(index: int) -> MonthReference:
    return MonthReference(year=index // 12, month=(index % 12) + 1)


def normalize_period_range(period_range: PeriodRange) -> PeriodRange:
    """Troca início/fim se o intervalo estiver invertido."""
    start_idx = to_month_index(period_range.start_year, period_range.start_month)
    end_idx = to_month_index(period_range.end_year, period_range.end_month)
    if start_idx <= end_idx:
        return period_range
    return PeriodRange(
        start_year=period_range.end_year,
        start_month=period_range.end_month,
        end_year=period_range.start_year,
        end_month=period_range.start_month,
    )


def is_within_period_range(year: int, month: int, period_range: PeriodRange | None) -> bool:
    """`True` sempre que `period_range` for `None` (nenhuma restrição = "Tudo")."""
    if period_range is None:
        return True
    normalized = normalize_period_range(period_range)
    idx = to_month_index(year, month)
    return (
        to_month_index(normalized.start_year, normalized.start_month)
        <= idx
        <= to_month_index(normalized.end_year, normalized.end_month)
    )


def enumerate_period_months(period_range: PeriodRange) -> list[MonthReference]:
    """Lista ordenada de {year, month} do início ao fim do intervalo, inclusive."""
    normalized = normalize_period_range(period_range)
    start_idx = to_month_index(normalized.start_year, normalized.start_month)
    end_idx = to_month_index(normalized.end_year, normalized.end_month)
    return [_from_month_index(idx) for idx in range(start_idx, end_idx + 1)]


def get_operational_period(year: int, month: int) -> OperationalPeriod:
    if not (1 <= month <= 12):
        raise ValueError("Competência mensal inválida.")
    if month == 12:
        return OperationalPeriod(period_year=year + 1, semester=Semester.S1)
    if month <= 5:
        return OperationalPeriod(period_year=year, semester=Semester.S1)
    return OperationalPeriod(period_year=year, semester=Semester.S2)


def cycle_from_year_semester(period_year: int, semester: Semester) -> PeriodRange:
    if semester == Semester.S2:
        return PeriodRange(
            start_year=period_year, start_month=6, end_year=period_year, end_month=11
        )
    return PeriodRange(
        start_year=period_year - 1, start_month=12, end_year=period_year, end_month=5
    )


def year_semester_from_cycle(cycle: PeriodRange) -> OperationalPeriod:
    semester = Semester.S1 if cycle.start_month == 12 else Semester.S2
    return OperationalPeriod(period_year=cycle.end_year, semester=semester)


def cycle_for_month(year: int, month: int) -> PeriodRange:
    op = get_operational_period(year, month)
    return cycle_from_year_semester(op.period_year, op.semester)


def get_current_cycle(reference: date | None = None) -> PeriodRange:
    ref = reference or date.today()
    return cycle_for_month(ref.year, ref.month)


def next_period(period_year: int, semester: Semester) -> OperationalPeriod:
    if semester == Semester.S1:
        return OperationalPeriod(period_year=period_year, semester=Semester.S2)
    return OperationalPeriod(period_year=period_year + 1, semester=Semester.S1)


def format_period_option_label(period_year: int, semester: Semester) -> str:
    cycle = cycle_from_year_semester(period_year, semester)
    start = f"{MONTH_NAMES[cycle.start_month - 1]}/{str(cycle.start_year)[-2:]}"
    end = f"{MONTH_NAMES[cycle.end_month - 1]}/{str(cycle.end_year)[-2:]}"
    return f"{period_year} {semester.value} · {start} - {end}"


def available_cycles_from_months(months: list[MonthReference]) -> list[PeriodRange]:
    """Somente ciclos com ao menos uma competência real, sem duplicidade, do mais recente ao mais antigo."""
    cycles: dict[tuple[int, int], PeriodRange] = {}
    for item in months:
        if not (1 <= item.month <= 12):
            continue
        cycle = cycle_for_month(item.year, item.month)
        cycles[(cycle.start_year, cycle.start_month)] = cycle
    return sorted(
        cycles.values(),
        key=lambda c: to_month_index(c.start_year, c.start_month),
        reverse=True,
    )


def format_period_range_label(
    period_range: PeriodRange | None, month_names: list[str] | None = None
) -> str:
    names = month_names or MONTH_NAMES
    if period_range is None:
        return "Tudo"
    normalized = normalize_period_range(period_range)
    start = f"{names[normalized.start_month - 1]}/{normalized.start_year}"
    end = f"{names[normalized.end_month - 1]}/{normalized.end_year}"
    return f"{start} – {end}"


def period_range_where(
    period_range: PeriodRange, year_field: str = "year", month_field: str = "month"
) -> tuple[str, dict[str, int]]:
    """Fragmento SQL (WHERE) equivalente a `is_within_period_range`.

    Retorna (sql_fragment, params) para uso com `text()` do SQLAlchemy.
    Inclusive nos dois extremos, atravessando a virada do ano civil quando
    necessário. Use nomes de parâmetro únicos ao compor múltiplas cláusulas.
    """
    normalized = normalize_period_range(period_range)
    sql = (
        f"(({year_field} > :period_start_year) "
        f"OR ({year_field} = :period_start_year AND {month_field} >= :period_start_month)) "
        f"AND (({year_field} < :period_end_year) "
        f"OR ({year_field} = :period_end_year AND {month_field} <= :period_end_month))"
    )
    params = {
        "period_start_year": normalized.start_year,
        "period_start_month": normalized.start_month,
        "period_end_year": normalized.end_year,
        "period_end_month": normalized.end_month,
    }
    return sql, params


def period_range_predicate(
    period_range: PeriodRange, year_col: Any, month_col: Any
) -> ColumnElement[bool]:
    """Equivalente ORM (SQLAlchemy ColumnElement) de `is_within_period_range`,
    para uso direto em `.where(...)` sem SQL textual."""
    from sqlalchemy import and_, or_

    normalized = normalize_period_range(period_range)
    return and_(
        or_(
            year_col > normalized.start_year,
            and_(year_col == normalized.start_year, month_col >= normalized.start_month),
        ),
        or_(
            year_col < normalized.end_year,
            and_(year_col == normalized.end_year, month_col <= normalized.end_month),
        ),
    )

File: frontend/utils/period.ts
import type { AvailablePeriod, PeriodRange } from "~/types/api";
import { MONTH_NAMES } from "~/utils/dates";

export const MONTHS = [
  "Janeiro", "Fevereiro", "Março", "Abril", "Maio", "Junho",
  "Julho", "Agosto", "Setembro", "Outubro", "Novembro", "Dezembro",
] as const;

export type Semester = "S1" | "S2";
export interface MonthReference { year: number; month: number }

/** Índice inteiro contínuo de um mês — compara/ordena intervalos entre anos. */
export function toMonthIndex(year: number, month: number): number {
  return year * 12 + (month - 1);
}

/** Troca início/fim se o intervalo estiver invertido. Porte de normalizePeriodRange. */
export function normalizePeriodRange(range: PeriodRange): PeriodRange {
  const startIdx = toMonthIndex(range.startYear, range.startMonth);
  const endIdx = toMonthIndex(range.endYear, range.endMonth);
  if (startIdx <= endIdx) return range;
  return { startYear: range.endYear, startMonth: range.endMonth, endYear: range.startYear, endMonth: range.startMonth };
}

/** `true` sempre que `range` for `null`/`undefined` (sem restrição = "Tudo"). */
export function isWithinPeriodRange(year: number, month: number, range: PeriodRange | null | undefined): boolean {
  if (!range) return true;
  const n = normalizePeriodRange(range);
  const idx = toMonthIndex(year, month);
  return idx >= toMonthIndex(n.startYear, n.startMonth) && idx <= toMonthIndex(n.endYear, n.endMonth);
}

/** Lista ordenada de {year, month} do início ao fim do intervalo, inclusive. */
export function enumeratePeriodMonths(range: PeriodRange): MonthReference[] {
  const n = normalizePeriodRange(range);
  const startIdx = toMonthIndex(n.startYear, n.startMonth);
  const endIdx = toMonthIndex(n.endYear, n.endMonth);
  const months: MonthReference[] = [];
  for (let idx = startIdx; idx <= endIdx; idx++) {
    months.push({ year: Math.floor(idx / 12), month: (((idx % 12) + 12) % 12) + 1 });
  }
  return months;
}

/** Classifica uma competência mensal no período operacional (Ano + Semestre).
 * Dezembro pertence ao S1 do ano SEGUINTE; janeiro-maio ao S1 do próprio ano;
 * junho-novembro ao S2 do próprio ano. Porte de getOperationalPeriod. */
export function getOperationalPeriod(year: number, month: number): { periodYear: number; semester: Semester } {
  if (!Number.isInteger(year) || !Number.isInteger(month) || month < 1 || month > 12) {
    throw new RangeError("Competência mensal inválida.");
  }
  if (month === 12) return { periodYear: year + 1, semester: "S1" };
  if (month <= 5) return { periodYear: year, semester: "S1" };
  return { periodYear: year, semester: "S2" };
}

/** Ano do período + Semestre -> intervalo de meses. Para S1, `periodYear` é
 * sempre o ano de TÉRMINO (maio). Porte de cycleFromYearSemester. */
export function cycleFromYearSemester(periodYear: number, semester: Semester): PeriodRange {
  return semester === "S2"
    ? { startYear: periodYear, startMonth: 6, endYear: periodYear, endMonth: 11 }
    : { startYear: periodYear - 1, startMonth: 12, endYear: periodYear, endMonth: 5 };
}

/** Inverso de cycleFromYearSemester. */
export function yearSemesterFromCycle(cycle: PeriodRange): { year: number; semester: Semester } {
  return { year: cycle.endYear, semester: cycle.startMonth === 12 ? "S1" : "S2" };
}

/** Ciclo operacional (intervalo de meses) que contém a competência informada. */
export function cycleForMonth(year: number, month: number): PeriodRange {
  const { periodYear, semester } = getOperationalPeriod(year, month);
  return cycleFromYearSemester(periodYear, semester);
}

/** Ciclo semestral ativo na data de referência — atalho "Ciclo atual". */
export function getCurrentCycle(reference: Date = new Date()): PeriodRange {
  return cycleForMonth(reference.getFullYear(), reference.getMonth() + 1);
}

/** Próximo semestre cronológico após Ano do período + Semestre informados. */
export function nextPeriod(periodYear: number, semester: Semester): { year: number; semester: Semester } {
  return semester === "S1" ? { year: periodYear, semester: "S2" } : { year: periodYear + 1, semester: "S1" };
}

/** Ex.: "2027 S1 · Dez/26 - Mai/27". */
export function formatPeriodOptionLabel(periodYear: number, semester: Semester): string {
  const cycle = cycleFromYearSemester(periodYear, semester);
  const start = `${MONTH_NAMES[cycle.startMonth - 1]}/${String(cycle.startYear).slice(-2)}`;
  const end = `${MONTH_NAMES[cycle.endMonth - 1]}/${String(cycle.endYear).slice(-2)}`;
  return `${periodYear} ${semester} · ${start} - ${end}`;
}

/** Ex.: "Dez/2026 – Mai/2027". Retorna "Tudo" quando `range` é `null`. */
export function formatPeriodRangeLabel(range: PeriodRange | null | undefined, monthNames: readonly string[] = MONTH_NAMES): string {
  if (!range) return "Tudo";
  const n = normalizePeriodRange(range);
  const start = `${monthNames[n.startMonth - 1] ?? n.startMonth}/${n.startYear}`;
  const end = `${monthNames[n.endMonth - 1] ?? n.endMonth}/${n.endYear}`;
  return `${start} – ${end}`;
}

export function operationalPeriod(referenceYear: number, semester: "S1" | "S2"): PeriodRange {
  return semester === "S1"
    ? { startYear: referenceYear - 1, startMonth: 12, endYear: referenceYear, endMonth: 5 }
    : { startYear: referenceYear, startMonth: 6, endYear: referenceYear, endMonth: 11 };
}

export function periodToQuery(period: PeriodRange | null | undefined): Record<string, number> {
  if (!period) return {};
  return {
    periodStartYear: period.startYear,
    periodStartMonth: period.startMonth,
    periodEndYear: period.endYear,
    periodEndMonth: period.endMonth,
  };
}

export function periodLabel(period: PeriodRange | null | undefined): string {
  if (!period) return "Todos os períodos";
  return `${MONTHS[period.startMonth - 1]}/${period.startYear} – ${MONTHS[period.endMonth - 1]}/${period.endYear}`;
}

export function periodFromAvailable(item: AvailablePeriod): PeriodRange {
  const first = item.competencies[0];
  const last = item.competencies.at(-1);
  return {
    startYear: first?.year ?? item.referenceYear,
    startMonth: first?.month ?? item.monthStart,
    endYear: last?.year ?? item.referenceYear,
    endMonth: last?.month ?? item.monthEnd,
  };
}

export function currentOperationalPeriod(date = new Date()): PeriodRange {
  const month = date.getMonth() + 1;
  const year = date.getFullYear();
  if (month >= 6 && month <= 11) return operationalPeriod(year, "S2");
  return operationalPeriod(month === 12 ? year + 1 : year, "S1");
}
